### PR TITLE
fix(cli): cross account asset upload no longer works

### PR DESCRIPTION
**Pull request by redbaron**
cdk_asset asset handlers use IAws to make calls to AWS APIs to discover
information about target environment: account id, region, partition.

Each asset is described by its manifest in a Cloud Assembly. This manifest
can contain placeholders to resolve by asset handlers when publishing
assets.

Previously `${Aws::Partition}` placeholder was derived from a code path used
to resolve `${Aws::AccountId}`, which was introducing a cyclic dependency for
cross account deployments:

- to replace partition placeholder it was assuming role in a target account
  to discover partition
- to assume role in a target account it needs to know full role ARN to assume
- role ARN contains partition placeholder

It was working for same account deployments and for non environment aware deployments,
because SdkProvider was always using current default (ambient) credentials without making
`AssumeRole` call, thus it was able to replace placeholders in asset manifest without
introducing a cyclic dependency.

To fix cross account deployments we introduce `IAWS.discoverPartition()` method to return
partition of default (ambient) credentials `cdk deploy` is called with. This works, because
cross partition `AssumeRole` calls are not possible, therefore it's enough to know our
default credentials partition.

Fixes https://github.com/aws/aws-cdk/issues/12151

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by rix0rrr** [comment]
You're a saint!

❤️

**Review by rix0rrr — DISMISSED** [review]

**Review by rix0rrr — APPROVED** [review]

File: packages/cdk-assets/lib/private/placeholders.ts [hunk 1]
```diff
@@ -22,7 +28,7 @@ export async function replaceAwsPlaceholders<A extends { region?: string }>(obje
       return (await account()).accountId;
     },
     async partition() {
-      return (await account()).partition;
+      return partition();
```
Comment (redbaron) [inline packages/cdk-assets/lib/private/placeholders.ts:31]: @rix0rrr , I think this is not right, doesn't it introduce infinite recursion?

Comment (redbaron) [reply packages/cdk-assets/lib/private/placeholders.ts:31]: apparently it is not

**Comment by redbaron** [comment]
@rix0rrr , validate-pr check is stuck :(

**Comment by redbaron** [comment]
@rix0rrr , you approved this PR to be merged, but it is stuck on `validate-pr` step. It might be broken because I briefly switched it to draft mode :(

**Comment by polothy** [comment]
Maybe rebase the branch to re-run the PR checks?

**Comment by redbaron** [comment]
@rix0rrr , would you mind to approve it again, only change is that I merged master to retrigger checks, because validate-pr was stuck

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**